test/jdk/javax/net/ssl/TLSv13/SSLEngineNoServerHelloClientShutdown.java

```diff
@@ -0,0 +1,305 @@
+        runDelegatedTasks(serverEngine);
+
+        cTOs.compact();
+
+        // server wrap
+        // produce server_hello
+        log("---Server Wrap server_hello---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // SH packet went missing.  Timeout on Client.
+
+        // server wrap
+        // produce other outbound messages
+        log("---Server Wrap---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // CCS packet went missing.  Timeout on Client.
+
+        // server wrap
+        // produce other outbound messages
+        log("---Server Wrap---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // EE/etc. packet went missing.  Timeout on Client.
+
+        // *Yawn*  No response.  Shutdown client
+        log("---Client closeOutbound---");
+        clientEngine.closeOutbound();
+
+        // Sends an unencrypted user_cancelled
+        log("---Client Wrap user_cancelled---");
+        clientResult = clientEngine.wrap(clientOut, cTOs);
+        logEngineStatus(clientEngine, clientResult);
+        runDelegatedTasks(clientEngine);
+
+        cTOs.flip();
+
+        // Server unwrap should process an unencrypted 2 byte packet,
+        log("---Server Unwrap user_cancelled alert---");
+        serverResult = serverEngine.unwrap(cTOs, serverIn);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+
+        cTOs.clear();
+
+        // Sends an unencrypted close_notify
+        log("---Client Wrap close_notify---");
+        clientResult = clientEngine.wrap(clientOut, cTOs);
+        logEngineStatus(clientEngine, clientResult);
+        assertEquals(clientResult.getStatus(), Status.CLOSED);
+        runDelegatedTasks(clientEngine);
+
+        cTOs.flip();
+
+        // Server unwrap should process an unencrypted 2 byte close_notify alert.
+        log("---Server Unwrap close_notify alert---");
+        serverResult = serverEngine.unwrap(cTOs, serverIn);
+        logEngineStatus(serverEngine, serverResult);
+        assertTrue(serverEngine.isInboundDone());
+        assertEquals(serverResult.getStatus(), Status.CLOSED);
+        runDelegatedTasks(serverEngine);
+
+        log("---Last Server Wrap ---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        assertTrue(serverEngine.isOutboundDone());
+        runDelegatedTasks(serverEngine);
+
+        /* TODO: Final client unwrap fails because server doesn't send an alert to terminate
+           the handshake after receiving close_notify alert from the client. Investigate why.
```

Comment:
Is this TODO item done?

Reply:
Not yet. I'm currently learning the internals of TLS code and going through TLS RFC in order to understand if it's an expected behavior or not. What do you think?

Reply:
you can't clear the sTOc buffer if you want this to work; TLS is running over TCP, and TCP never loses data. So, the final client unwrap is expected to deal with serverHello and all handshake messages before receiving the encrypted close_notify alert.

Reply:
@djelinski I see, thanks! But here is the exception I'm getting, it's not about unexpected command. Can we get it because server is using TLSv1.3 and client is using TLSv1.2? How do we switch server to TLSv1.2 after encontering this situation?
`---Last Client Unwrap---
javax.net.ssl.SSLException: Unrecognized SSL message, plaintext connection?
	at java.base/sun.security.ssl.SSLEngineInputRecord.bytesInCompletePacket(SSLEngineInputRecord.java:145)
	at java.base/sun.security.ssl.SSLEngineInputRecord.bytesInCompletePacket(SSLEngineInputRecord.java:64)
	at java.base/sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:613)
	at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:507)
	at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:483)
	at java.base/javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:679)
	at SSLEngineNoServerHelloClientShutdown.runTestUserCancelled(SSLEngineNoServerHelloClientShutdown.java:244)
	at SSLEngineNoServerHelloClientShutdown.main(SSLEngineNoServerHelloClientShutdown.java:140)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:573)
`

Reply:
Yes, I figured out how to switch the server to plaintext and now client can read the last server message. Now it produces `(unexpected_message) Received close_notify during handshake` error. Client expects `user_cancelled` alert first before getting `close_notify` during handshake.

Reply:
Fixed that client issue by requiring to always send `user_cancelled` alert first during handshake, will be pushing a new commit soon.